leshan-core/src/main/java/org/eclipse/leshan/core/util/X509CertUtil.java

```diff
@@ -0,0 +1,310 @@
+                    value = new StringBuilder();
+                } else {
+                    value.append(ch);
+                }
+            }
+        }
+
+        if (key.length() > 0) {
+            map.put(key.toString(), value.toString());
+        }
+
+        return map;
+    }
+
+    /**
+     * Extracts field from given principal's name.
+     * <p>
+     * Notes: Can parse most name strings but HEX encoded DER values are returned back in DER HEX form.
+     * Notes: Only understands X500Principal.
+     *
+     * @param principal Source principal.
+     * @param field     Field key as defined in RFC 2253.
+     * @return null or value as string.
+     */
+    public static String getPrincipalField(Principal principal, String field) {
+        if (principal instanceof X500Principal) {
+            X500Principal x500Principal = (X500Principal) principal;
+
+            // Extra practical OID that Java implementation hides for "not being standard" even thou
+            // list in RFC is only example OIDs list (see JDK class AVAKeyword). These extra OID's are
+            // actually defined in Sun packages but we should not access them as they are not standard
+            // Java classes.
+            HashMap<String, String> extraOids = new HashMap<String, String>();
+            extraOids.put("2.5.4.5", "SERIALNUMBER");
+
+            String dn = x500Principal.getName(X500Principal.RFC2253, extraOids);
+
+            Map<String, String> fields = parseRfc2253Name(dn);
+
+            if (fields.containsKey(field)) {
+                return fields.get(field);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * DNS name matcher
+     * <p>
+     * Supports matching with wildcard DNS names.
+     *
+     * @param matcher Matcher pattern
+     * @param target  Target DNS name to check
+     * @return true if matches, false otherwise
+     */
+    private static boolean dnsNameMatch(String matcher, String target) {
+        if (matcher.startsWith("*.")) {
+            // Simple filtering out
+            if (!target.endsWith(matcher.substring(1)))
+                return false;
+
+            // Wildcards only work in one sub level so no extra dots
+            String host = target.substring(0, target.length() - (matcher.length() - 1));
+            return host.indexOf('.') == -1;
+        }
+        return matcher.equals(target);
+    }
+
+    /**
+     * Match DNS name against X.509 Certificate's Subjects
+     *
+     * @param certificate Target X.509 certificate
+     * @param dnsName     DNS name to match
+     * @return True if match, false otherwise
+     */
+    public static boolean matchSubjectDnsName(X509Certificate certificate, String dnsName) {
+        try {
+            // First one need to check SANs if they are present
+
+            final Collection<List<?>> sans = certificate.getSubjectAlternativeNames();
+
+            if (sans != null) {
+                for (final List<?> san : sans) {
```

Comment:
Generally we avoid final key word for variable or parameter if this is not needed.
Any reason you use it ?

Reply:
Will remove. Probably habit from some other programming language.